##Basic usage

The program is command-line driven. The Syntax is
   
    $ gedi [ -options | --options parameters ]

where options preceded by a single hyphen work by themselves, and those with two hyphens must be followed by a parameter.

###IL/CL

There are two main options specifying the analysis class (either one must be present):

    -il    Independent loci (IL) analysis. Each SNP is treated independent of others.
    -cl    Collective loci  (CL) analysis. SNPs interact with one another.

These two analysis options are described in detail in [IL](il.md) and [CL](cl.md).

###Output file

By default, in addition to writing messages to the standard output, the program will write results to a file called "gedi.out" (overwritten it it already exists in the working directory).
To specify an output file, use

    --out output_file     Writes the results to "output_file" (default: "gedi.out")

###Models

Three models are supported for allelic effects (a=major allele, A=minor allele):

    -dominant    Dominant model  (aa=no effect, Aa/AA=same effect); default
    -recessive   Recessive model (aa/Aa=no effect, AA=effect)
    -genotypic   Genotypic model (aa=no effect, Aa=effect 1, AA=effect 2)

###File formats

####1. ASCII files

  The input data need to be in SNP-major format. The ASCII version is the PLINK "transposed" format, which consists of a phenotype file (e.g., "file.tfam"), where each line contains six entries for one individual:

    Family_ID  Ind_ID   Pat_ID   Mat_ID Sex Phenotype

  For example,

    857688     857688   0        0      1   2
    866290     866290   0        0      2   1

  See PLINK web [site](http://pngu.mgh.harvard.edu/~purcell/plink/) for more details. In GeDI, only phenotypes are used (coded 1=control, 2=case). The file "file.tped" lists genotypes of each SNP per line:

    Chr#   SNP_ID    dist   pos(bp)  Ind1_A1 Ind1_A2  ... indN_A1 indN_A2

  For example,

    1      rs311560  0      743268   T T  G T  ...  T T
    1      rs121219  0      766409   G A  A A  ...  A A

  The two files have to be consistent. It is required that alleles are coded as T, C, A, G. Other letters will be interpreted as "missing" These files are specified using

    --tped file.tped --tfam file.tfam

  By default, major/minor alleles will be defined with respect to the allele frequencies over the pooled (case and control combined) sample. Alternatively, they can be defined with respect to control group only by specifying

    -ma_ctl

  Data files in other formats can be converted into tped/tfam files using [PLINK](https://www.cog-genomics.org/plink2) (--recode transpose).

####2. Binary files

The tped file can be too big for whole genome data; GeDI can read PLINK binary files, which come in bundles of three files: "file.bed", "file.bim", and "file.fam", each storing genotypes, SNP information, and phenotypes, respectively. They are generated by the [PLINK](https://www.cog-genomics.org/plink2) "--make-bed" option. Read these files into GeDI with

    --bfile file

and it will look for the three files at the working directory. An error will occur when any of them are missing.

***
[Up](README.md)

[Next](il.md)
